### PR TITLE
Initial version

### LINKED ISSUE #1: Initial functioning
Things to fix before first merge (and demo Friday):

- [x]  get react Column layout working
- [x] clean up debug text
- [x] display watershed names in dropdown
- [x] format data display better somehow
- [ ] allow variable choice
- [x] maybe grab an ncWMS layer?
- [x] get Rod to check hooks and state
- [ ] basic README and documentation

### PR DESCRIPTION
**Pull request by corviday**
Much still needs to be done, but this is an early check, to get some eyes on the code. Unfortunately, there are no tests or documentation yet.

Specific questions:

* why isn't the layout working?
* are the React hooks being using properly?

resolves #1

### CODE REVIEW

File: src/components/AnnualCycleGraph/AnnualCycleGraph.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
+import React from 'react';
+import {values} from 'lodash';
+
+
+//this weird sequence avoids crashing npm when plotly is loaded.
```
Comment (corviday) [inline src/components/AnnualCycleGraph/AnnualCycleGraph.js:5]: need to link to a better description of the issue.

File: src/components/AnnualCycleGraph/AnnualCycleGraph.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+                    mode: 'lines+markers',
+                    marker: {color: 'red'},
+                },
+            ]}
+            layout={ {width: 500, height: 500, title: 'Mean Daily Maximum Temperature'} }
+        />
+      );
+}
+
+export default AnnualCycleGraph;
```
Comment (corviday) [inline src/components/AnnualCycleGraph/AnnualCycleGraph.js:43]: Need better labels, ability to choose a specific dataset.

File: src/components/App/App.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+              <DataController
+                currentRegionBoundary={currentRegionBoundary}
+              />
+            </Col> 
+          </Row>
+        </Container>        
+      </header>
+    </div>
+  );
+}
```
Comment (corviday) [inline src/components/App/App.js:46]: Container/Row/Column isn't working. All components are stacked vertically. Any suggestions, @rod-glover ? Do I need something in the CSS to make them work?

File: src/components/AreaController/AreaController.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+import {findIndex} from 'lodash';
+
+function AreaController({onChangeRegionName, onChangeRegionBoundary}) {
+
+  const [regionNames, setRegionNames] = useState([]);
+  const [regionBoundaries, setRegionBoundaries] = useState([]);
+  const [currentRegionName, setCurrentRegionName] = useState("No Region Selected");
+  const [currentRegionBoundary, setCurrentRegionBoundary] = useState(null);
+  const [regionAreas, setRegionAreas] = useState();
+  const [currentRegionArea, setCurrentRegionArea] = useState();
```
Comment (corviday) [inline src/components/AreaController/AreaController.js:13]: Need to bundle region boundaries, area, and names into a single data structure, especially as we want to add a lot more region metadata,

File: src/components/DataController/DataController.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+import React, {useState} from 'react';
+import moment from 'moment/moment';
+
+
+function DataController({currentRegionBoundary}) {
+  
+  const [monthlyTimeSeries, setMonthlyTimeSeries] = useState(null);
+  const [prevRegion, setPrevRegion] = useState(null);
+  
+  // fetch data and format it  - currently just displaying as text.
```
Comment (corviday) [inline src/components/DataController/DataController.js:12]: whoops, that's not true. Data is displayed as an extremely primitive graph.

File: src/components/DataMap/DataMap.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  return (
+    <div className="DataMap">
+        <BCBaseMap
+          id={"map"}
+          zoom={viewport.zoom}
+          center={viewport.center}
+        >
+          <SetView view={viewport}/>
+          <SimpleGeoJSON data={boundaryFeature} fill={false}/>
+          <WMSTileLayer
```
Comment (corviday) [inline src/components/DataMap/DataMap.js:23]: This was hardwired for the demo, but will eventually be selectable.

File: src/components/SimpleGeoJSON/SimpleGeoJSON.js [hunk 7]
```diff
@@ -0,0 +1,71 @@
+//Copy-pastedfrom plan2adapt - should add to pcic-react-leaflet instead!
```
Comment (corviday) [inline src/components/SimpleGeoJSON/SimpleGeoJSON.js:1]: Apparently leaflet now has this functionalitty natively - update to that.

File: src/components/App/App.css [hunk 8]
```diff
@@ -0,0 +1,38 @@
+.App {
```
Comment (rod-glover) [inline src/components/App/App.css:1]: Pretty sure most of this is cruft left over from the default app that create-react-app gives you.

File: src/components/App/App.js [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+import {Container, Row, Col} from 'react-bootstrap';
+import DataController from '../DataController/DataController.js'
+import MapController from '../MapController/MapController.js'
+import AreaController from '../AreaController/AreaController.js'
+import React, {useState} from 'react'
+
+function App() {
+  const [currentRegionName, setCurrentRegionName] = useState(null);
+  const [currentRegionBoundary, setCurrentRegionBoundary] = useState(null);
```
Comment (rod-glover) [inline src/components/App/App.js:11]: See comment about coalescing state into a single object.

File: src/components/AreaController/AreaController.js (hunk 4 above)
Comment (rod-glover) [reply src/components/AreaController/AreaController.js:13]: Totally agree. See my global comment in this review.

File: src/components/AreaController/AreaController.js [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+  //fetch region list from geoserver if we don't already have it.
```
Comment (rod-glover) [inline src/components/AreaController/AreaController.js:15]: See global comment about `useEffect`

File: src/components/AreaSelector/AreaSelector.js [hunk 11]
```diff
@@ -0,0 +1,24 @@
+import {SelectWithValueReplacement as Select} from 'pcic-react-components'
+import React from 'react'
+
+function AreaSelector({regionNames, onChange, currentRegion}) {
+  var valueOptions = [];
+  regionNames.forEach(name => valueOptions.push({"value": name, "label": name}));
```
Comment (rod-glover) [inline src/components/AreaSelector/AreaSelector.js:6]: `_.map` would do this without explicitly constructing and pushing onto an initially empty array. Shorter, fewer oppos for mistakes.

File: src/components/DataController/DataController.js [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+    setPrevRegion(currentRegionBoundary);
+  }
+  
+  function monthlyTimeseriesText() {
+      //test function that just dispays data as text.
+      if(monthlyTimeSeries === null) {
+          return "No data available";
+      }
+      else {
+        var dataStrings = [];
```
Comment (rod-glover) [inline src/components/DataController/DataController.js:27]: Ditto comment about `_.map`

**Review by rod-glover — COMMENTED** [review]
Main/global observations:

### State patterns

The general rule for state is ["lift state up"](https://reactjs.org/docs/lifting-state-up.html). Additionally, or a corollary of this, is that state should never be duplicated nor should it duplicate properties.

`AreaController` duplicates state in that it holds the user's region selection in state variables, and also communicates those by a callback to `App`, which also stores those values in its own state variables. So state is being duplicated. I think `AreaController` doesn't need to and shouldn't hold this state: instead it should accept the state from `App` as props, and communicate it back to its caller (`App`) via callbacks as you have done. But no internal copy.

### Handling region info 

I completely agree with you that the region info should all be in a single data item. It looks like you are unpacking an awkward data structure (into 3). Instead you could restructure it into a single data structure, or you could create accessor functions that make getting the desired parts out of a raw region item simpler (and DRY up the access wherever it occurs).

Your choice to segregate the fetching of region data into `AppController` is interesting. If you will never need the list of regions for any other use, it's nice in that it hides some complexity from the main app. But you could do the fetching in a parent component (`App`) and pass it in to the selector component. If you do that, plus lifting the selection state up, you can get rid of `AppController` altogether and just have `AppSelector`.

### Usage of React hooks (fetching pattern)

All side effects in a component that uses React hooks must be performed inside a [`useEffect` hook](https://reactjs.org/docs/hooks-effect.html). Updating state from an asynchronous fetch is a classic side effect. The pattern is

```js
function MyComponent() {
  const [things, setThings] = useState(null);

  useEffect(() => {
    fetchThings.then(setThings);
  }, []);  // see note below

  // Do stuff that uses `things`.
  // ...
}
```

Apart from its required use for side effects, `useEffect` offers some extras. One of them is keeping track of whether anything changed in the list of values you supply as a second argument and executing when and only when they have. Empty list means do it once, which is typical for fetching a fixed set of data like the regions.

This applies in two places in the code:

- `AreaController`: Call to `fetchWatersheds` and subsequent state setting should be in a `useEffect`. Use empty array for second arg because it need only be done once.
- `DataController`: Call to `testDataRequest` and subsequent state setting should be done in a `useEffect`. There's no need to keep a `prevRegion` or explicitly test it; instead pass `currentRegionBoundary` in the second argument. When its value changes (shallow comparison), the function passed to `useEffect` is called again. 

### Naming tweaks

"Controller" is redundant in the name of a component. All components that maintain state and pass it into subcomponents are controllers, or rather components that receive values from outside and pass updates to that value back (via callback) to the calling component are *controlled components*. Calling a component `XController` is kind of like calling a variable `countInteger`. Instead we can name components more explicitly for what they are.

- `MapController`: `MapDisplay`  ... also why does this component exist? It doesn't add anything to `DataMap`. Eliminate? Or will more go in here?
- `AreaController`: This component could go away (`App` becomes the state owner), so we don't have to find a better name for it. The fetch in it can go either in `AreaSelector` or in `App`.
- `DataController`: `DataDisplay`? `DataGraphs`?

### Remove disused things

- React logos + entries in manifest for them
- CRA cruft
- other?

File: src/components/AnnualCycleGraph/AnnualCycleGraph.js (hunk 2 above)
Comment (corviday) [reply src/components/AnnualCycleGraph/AnnualCycleGraph.js:61]: Better labels were done by @helfy18 ; I'm going to bump data selection to a later PR, as we've been working on map and watershed interfaces.

**Pull request merged**